File: src/application/services/triage.service.ts
import { MedicalProtocolRepository } from '../../infrastructure/repositories/medical-protocol.repository';
import { CaseReportRepository } from '../../infrastructure/repositories/case-report.repository';
import { CaseReportService } from './case-report.service';
import { UrgencyLevel } from '@prisma/client';

interface TriageResult {
    urgencyLevel: UrgencyLevel;
    matchedProtocol?: {
        id: string;
        name: string;
        steps: string;
    };
    keywords: string[];
}

export class TriageService {
    private protocolRepo = new MedicalProtocolRepository();
    private caseReportService = new CaseReportService(new CaseReportRepository());

    /**
     * Analyze SMS content and determine urgency level
     */
    async analyzeSMS(content: string): Promise<TriageResult> {
        const normalizedContent = content.toLowerCase();
        const words = normalizedContent.split(/\s+/);

        // Get all active protocols
        const protocols = await this.protocolRepo.findAll();

        let bestMatch: any = null;
        let maxMatches = 0;
        const matchedKeywords: string[] = [];

        // Find protocol with most keyword matches
        for (const protocol of protocols) {
            if (!protocol.isActive) continue;

            let matches = 0;
            const protocolKeywords: string[] = [];

            for (const keyword of protocol.keywords) {
                const keywordLower = keyword.toLowerCase();
                if (normalizedContent.includes(keywordLower)) {
                    matches++;
                    protocolKeywords.push(keyword);
                }
            }

            if (matches > maxMatches) {
                maxMatches = matches;
                bestMatch = protocol;
                matchedKeywords.length = 0;
                matchedKeywords.push(...protocolKeywords);
            }
        }

        // Determine urgency level
        let urgencyLevel: UrgencyLevel = 'NORMAL';

        if (bestMatch) {
            urgencyLevel = bestMatch.urgencyLevel;
        } else {
            // Fallback: check for critical keywords
            const criticalKeywords = ['urgence', 'critique', 'grave', 'mourant', 'inconscient', 'convulsion'];
            const urgentKeywords = ['fièvre', 'douleur', 'saignement', 'vomissement'];

            if (criticalKeywords.some(kw => normalizedContent.includes(kw))) {
                urgencyLevel = 'CRITICAL';
            } else if (urgentKeywords.some(kw => normalizedContent.includes(kw))) {
                urgencyLevel = 'URGENT';
            }
        }

        return {
            urgencyLevel,
            matchedProtocol: bestMatch ? {
                id: bestMatch.id,
                name: bestMatch.name,
                steps: bestMatch.steps,
            } : undefined,
            keywords: matchedKeywords,
        };
    }

    /**
     * Create CaseReport from SMS triage
     */
    async createCaseReportFromSMS(
        agentId: string,
        patientId: string | undefined,
        symptoms: string,
        triageResult: TriageResult
    ) {
        return this.caseReportService.createCaseReport({
            agentId,
            patientId,
            symptoms,
            urgency: triageResult.urgencyLevel,
            channel: 'SMS',
        });
    }

    /**
     * Generate automatic response based on triage
     */
    generateResponse(triageResult: TriageResult, caseReportId: string): string {
        const { urgencyLevel, matchedProtocol } = triageResult;

        let response = `Cas enregistré (ID: ${caseReportId.substring(0, 8)}).\n`;

        if (matchedProtocol) {
            response += `Protocole: ${matchedProtocol.name}\n`;

            // Parse steps and extract key instructions
            try {
                const steps = JSON.parse(matchedProtocol.steps);
                if (steps.traitement) {
                    response += `Traitement: ${steps.traitement.substring(0, 100)}...\n`;
                }
            } catch (e) {
                // If steps is not JSON, skip
            }
        }

        if (urgencyLevel === 'CRITICAL') {
            response += '⚠️ URGENCE CRITIQUE - Médecin notifié. Référer immédiatement.';
        } else if (urgencyLevel === 'URGENT') {
            response += '⚠️ URGENT - Médecin sera consulté rapidement.';
        } else {
            response += 'Cas en attente de consultation.';
        }

        return response;
    }
}

File: sch_mobile/lib/features/triage/domain/services/triage_service.dart
import 'package:dio/dio.dart';
import '../../data/models/case_report_model.dart';
import '../../data/models/medical_protocol_model.dart';
import '../../data/models/triage_result_model.dart';
import '../../data/repositories/medical_protocol_repository.dart';

/// Service for analyzing symptoms and determining urgency level
class TriageService {
  final MedicalProtocolRepository _protocolRepo;
  final Dio _dio;

  TriageService(this._protocolRepo, this._dio);

  /// Analyze symptoms text and determine urgency level
  /// 
  /// This method:
  /// 1. Normalizes the input text
  /// 2. Retrieves active medical protocols
  /// 3. Finds the best matching protocol based on keywords
  /// 4. Determines urgency level
  /// 5. Generates recommendations
  Future<TriageResultModel> analyzeSymptoms(String symptoms) async {
    // 1. Normalize text
    final normalized = symptoms.toLowerCase().trim();
    if (normalized.isEmpty) {
      return TriageResultModel(
        urgencyLevel: UrgencyLevel.normal,
        keywords: [],
        recommendation: 'Veuillez décrire les symptômes',
      );
    }

    final words = normalized.split(RegExp(r'\s+'));

    // 2. Get active protocols
    final protocols = await _protocolRepo.getActiveProtocols();

    // 3. Find best match
    MedicalProtocolModel? bestMatch;
    int maxMatches = 0;
    List<String> matchedKeywords = [];

    for (final protocol in protocols) {
      int matches = 0;
      List<String> protocolKeywords = [];

      for (final keyword in protocol.keywords) {
        final keywordLower = keyword.toLowerCase();
        if (normalized.contains(keywordLower)) {
          matches++;
          protocolKeywords.add(keyword);
        }
      }

      if (matches > maxMatches) {
        maxMatches = matches;
        bestMatch = protocol;
        matchedKeywords = protocolKeywords;
      }
    }

    // 4. Try Online AI Triage first, fallback to offline
    try {
      final response = await _dio.post('/triage/analyze', data: {
        'symptoms': symptoms,
      });

      if (response.data != null && response.data['status'] == 'success') {
        final aiData = response.data['data'];
        
        // Parse the AI Urgency string to Enum
        UrgencyLevel aiUrgency = switch (aiData['urgency']) {
          'CRITICAL' => UrgencyLevel.critical,
          'URGENT' => UrgencyLevel.urgent,
          _ => UrgencyLevel.normal,
        };

        return TriageResultModel(
          urgencyLevel: aiUrgency,
          matchedProtocol: bestMatch,
          keywords: (aiData['keywords'] as List).cast<String>(),
          recommendation: aiData['clinicalAssessment'], // Use AI Assessment as recommendation
        );
      }
    } catch (e) {
      print('⚠️ Online AI Triage failed, falling back to offline algorithm: $e');
    }

    // 5. Offline Fallback: Determine urgency level
    UrgencyLevel urgency = UrgencyLevel.normal;

    if (bestMatch != null) {
      urgency = bestMatch.urgencyLevel;
    } else {
      // Fallback: check for critical keywords
      urgency = _determineUrgencyFromKeywords(normalized);
    }

    // 6. Generate fallback recommendation
    final recommendation = _generateRecommendation(urgency, bestMatch);

    return TriageResultModel(
      urgencyLevel: urgency,
      matchedProtocol: bestMatch,
      keywords: matchedKeywords.isEmpty ? ['analyse', 'hors-ligne'] : matchedKeywords,
      recommendation: recommendation,
    );
  }

  /// Determine urgency from critical/urgent keywords when no protocol matches
  UrgencyLevel _determineUrgencyFromKeywords(String normalizedText) {
    // Critical keywords (French and Créole)
    final criticalKeywords = [
      'urgence',
      'critique',
      'grave',
      'mourant',
      'inconscient',
      'convulsion',
      'kriz', // crise (Créole)
      'mouri', // mourir (Créole)
      'san', // sang (Créole)
      'pa respire', // ne respire pas (Créole)
    ];

    // Urgent keywords (French and Créole)
    final urgentKeywords = [
      'fièvre',
      'douleur',
      'saignement',
      'vomissement',
      'diarrhée',
      'lafyèv', // fièvre (Créole)
      'doule', // douleur (Créole)
      'vomi', // vomissement (Créole)
      'dyare', // diarrhée (Créole)
    ];

    if (criticalKeywords.any((kw) => normalizedText.contains(kw))) {
      return UrgencyLevel.critical;
    } else if (urgentKeywords.any((kw) => normalizedText.contains(kw))) {
      return UrgencyLevel.urgent;
    }

    return UrgencyLevel.normal;
  }

  /// Generate recommendation based on urgency and protocol
  String _generateRecommendation(
    UrgencyLevel urgency,
    MedicalProtocolModel? protocol,
  ) {
    if (protocol != null) {
      return 'Protocole recommandé: ${protocol.displayName}. '
          'Veuillez suivre les étapes du protocole.';
    }

    return switch (urgency) {
      UrgencyLevel.critical =>
        '🚨 URGENCE CRITIQUE: Référer immédiatement vers un hôpital. '
            'Le patient nécessite une attention médicale urgente.',
      UrgencyLevel.urgent =>
        '⚠️ URGENT: Consulter un médecin dans les plus brefs délais. '
            'Surveillance étroite recommandée.',
      UrgencyLevel.normal =>
        'ℹ️ Cas normal: Continuer la surveillance. '
            'Consulter si les symptômes s\'aggravent.',
    };
  }
}
